brave/src/test/java/brave/sampler/RateLimitingSamplerSoakTest.java

```diff
@@ -0,0 +1,66 @@
+package brave.sampler;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import org.junit.experimental.theories.DataPoints;
+import org.junit.experimental.theories.Theories;
+import org.junit.experimental.theories.Theory;
+import org.junit.runner.RunWith;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assumptions.assumeThat;
+
+@RunWith(Theories.class)
+public class RateLimitingSamplerSoakTest {
+
+  @DataPoints public static final int[] SAMPLE_RESERVOIRS = {11, 101, 1001, 1_000_001};
+
+  /** This test will take a little over a second per reservoir */
+  @Theory public void retainsPerSampleRate(int reservoir) throws Exception {
+    Sampler sampler = RateLimitingSampler.create(reservoir);
+
+    // We want to make sure we fill up the entire second, so
+    long startTick = System.nanoTime();
+    long lastDecisecond = startTick + TimeUnit.MILLISECONDS.toNanos(990);
+    // Because exacts don't work with Thread.sleep + operation overhead, set a deadline just under a second
+    long deadline = startTick + TimeUnit.MILLISECONDS.toNanos(998);
+
+    AtomicBoolean hitLastDecisecond = new AtomicBoolean();
+
+    AtomicLong passed = new AtomicLong();
+    Runnable sample = () -> {
+      long tick = System.nanoTime();
+      if (tick > deadline) return;
+      if (tick >= lastDecisecond) hitLastDecisecond.set(true);
+
+      if (sampler.isSampled(0L)) passed.incrementAndGet();
+    };
+
+    ExecutorService service = Executors.newFixedThreadPool(10);
+    List<Future<?>> futures = new ArrayList<>();
+    do {
+      if (reservoir > 10) {  // execute one tenth of our reservoir
+        for (int i = 0; i < reservoir / 10; i++) futures.add(service.submit(sample));
+      } else {// don't divide by 10!
+        service.submit(sample);
+      }
+
+      for (Future<?> future : futures) future.get();
+      futures.clear(); // block instead of filling up the executor
+      Thread.sleep(9);
+    } while (System.nanoTime() < deadline);
+    service.shutdown();
+    service.awaitTermination(1, TimeUnit.SECONDS);
+
+    assertThat(passed.get()).isEqualTo(reservoir);
```

Comment:
I'm surprised it's possible to have an exact match here given timing issues due to thread scheduling. Is it ok to do an exact match?

Reply:
this is why the threads skip out if they notice past a deadline